**Understanding the dataset**

In [37]:
#import necessary libraries
import pandas as pd
import numpy as np

from IPython.display import display

In [ ]:
#read excel data 
df = pd.read_excel('../data/raw/E Commerce Dataset.xlsx', sheet_name='E Comm')

#display 5 first rows
df.head(20)

,CustomerID,Churn,Tenure,PreferredLoginDevice,CityTier,WarehouseToHome,PreferredPaymentMode,Gender,HourSpendOnApp,NumberOfDeviceRegistered,PreferedOrderCat,SatisfactionScore,MaritalStatus,NumberOfAddress,Complain,OrderAmountHikeFromlastYear,CouponUsed,OrderCount,DaySinceLastOrder,CashbackAmount
0,50001,1,4.0,Mobile Phone,3,6.0,Debit Card,Female,3.0,3,Laptop & Accessory,2,Single,9,1,11.0,1.0,1.0,5.0,159.93
1,50002,1,NaN,Phone,1,8.0,UPI,Male,3.0,4,Mobile,3,Single,7,1,15.0,0.0,1.0,0.0,120.90
2,50003,1,NaN,Phone,1,30.0,Debit Card,Male,2.0,4,Mobile,3,Single,6,1,14.0,0.0,1.0,3.0,120.28
3,50004,1,0.0,Phone,3,15.0,Debit Card,Male,2.0,4,Laptop & Accessory,5,Single,8,0,23.0,0.0,1.0,3.0,134.07
4,50005,1,0.0,Phone,1,12.0,CC,Male,NaN,3,Mobile,5,Single,3,0,11.0,1.0,1.0,3.0,129.60
5,50006,1,0.0,Computer,1,22.0,Debit Card,Female,3.0,5,Mobile Phone,5,Single,2,1,22.0,4.0,6.0,7.0,139.19
6,50007,1,NaN,Phone,3,11.0,Cash on Delivery,Male,2.0,3,Laptop & Accessory,2,Divorced,4,0,14.0,0.0,1.0,0.0,120.86
7,50008,1,NaN,Phone,1,6.0,CC,Male,3.0,3,Mobile,2,Divorced,3,1,16.0,2.0,2.0,0.0,122.93
8,50009,1,13.0,Phone,3,9.0,E wallet,Male,NaN,4,Mobile,3,Divorced,2,1,14.0,0.0,1.0,2.0,126.83
9,50010,1,NaN,Phone,1,31.0,Debit Card,Male,2.0,5,Mobile,3,Single,2,0,12.0,1.0,1.0,1.0,122.93


In [39]:
#check the data in general
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   CustomerID                   5630 non-null   int64  
 1   Churn                        5630 non-null   int64  
 2   Tenure                       5366 non-null   float64
 3   PreferredLoginDevice         5630 non-null   str    
 4   CityTier                     5630 non-null   int64  
 5   WarehouseToHome              5379 non-null   float64
 6   PreferredPaymentMode         5630 non-null   str    
 7   Gender                       5630 non-null   str    
 8   HourSpendOnApp               5375 non-null   float64
 9   NumberOfDeviceRegistered     5630 non-null   int64  
 10  PreferedOrderCat             5630 non-null   str    
 11  SatisfactionScore            5630 non-null   int64  
 12  MaritalStatus                5630 non-null   str    
 13  NumberOfAddress              

In [40]:
#checking in detail which columns has missing values  
missing = df.isnull().sum()
missing[missing > 0]

Tenure                         264
WarehouseToHome                251
HourSpendOnApp                 255
OrderAmountHikeFromlastYear    265
CouponUsed                     256
OrderCount                     258
DaySinceLastOrder              307
dtype: int64

In [41]:
#check duplicate
df.duplicated().sum()

np.int64(0)

In [47]:
#check inconsistency
categorical_cols = df.select_dtypes(exclude=['number']).columns

for col in categorical_cols:
    display(df[col].value_counts())

PreferredLoginDevice
Mobile Phone    2765
Computer        1634
Phone           1231
Name: count, dtype: int64

PreferredPaymentMode
Debit Card          2314
Credit Card         1501
E wallet             614
UPI                  414
COD                  365
CC                   273
Cash on Delivery     149
Name: count, dtype: int64

Gender
Male      3384
Female    2246
Name: count, dtype: int64

PreferedOrderCat
Laptop & Accessory    2050
Mobile Phone          1271
Fashion                826
Mobile                 809
Grocery                410
Others                 264
Name: count, dtype: int64

MaritalStatus
Married     2986
Single      1796
Divorced     848
Name: count, dtype: int64

In [48]:
#check dtypes
df.dtypes

CustomerID                       int64
Churn                            int64
Tenure                         float64
PreferredLoginDevice               str
CityTier                         int64
WarehouseToHome                float64
PreferredPaymentMode               str
Gender                             str
HourSpendOnApp                 float64
NumberOfDeviceRegistered         int64
PreferedOrderCat                   str
SatisfactionScore                int64
MaritalStatus                      str
NumberOfAddress                  int64
Complain                         int64
OrderAmountHikeFromlastYear    float64
CouponUsed                     float64
OrderCount                     float64
DaySinceLastOrder              float64
CashbackAmount                 float64
dtype: object

In [43]:
#check statistic view
df.describe().transpose()

,count,mean,std,min,25%,50%,75%,max
CustomerID,5630.0,52815.500000,1625.385339,50001.0,51408.25,52815.50,54222.7500,55630.00
Churn,5630.0,0.168384,0.374240,0.0,0.00,0.00,0.0000,1.00
Tenure,5366.0,10.189899,8.557241,0.0,2.00,9.00,16.0000,61.00
CityTier,5630.0,1.654707,0.915389,1.0,1.00,1.00,3.0000,3.00
WarehouseToHome,5379.0,15.639896,8.531475,5.0,9.00,14.00,20.0000,127.00
HourSpendOnApp,5375.0,2.931535,0.721926,0.0,2.00,3.00,3.0000,5.00
NumberOfDeviceRegistered,5630.0,3.688988,1.023999,1.0,3.00,4.00,4.0000,6.00
SatisfactionScore,5630.0,3.066785,1.380194,1.0,2.00,3.00,4.0000,5.00
NumberOfAddress,5630.0,4.214032,2.583586,1.0,2.00,3.00,6.0000,22.00
Complain,5630.0,0.284902,0.451408,0.0,0.00,0.00,1.0000,1.00


***Summary***:
- 5,630 rows and 20 columns (8 float, 7 integer, 5 categorical/string variables)
- 0 duplicate rows but found inconsistency
- DaySinceLastOrder (307 missing)
- OrderAmountHikeFromlastYear (265 missing)
- Tenure (264 missing)
- OrderCount (258 missing)
- CouponUsed (256 missing)
- HourSpendOnApp (255 missing)
- WarehouseToHome (251 missing)


- PreferredLoginDevice:	Mobile Phone (2,765) and Phone (1,231) represent the same category
- PreferredPaymentMode:CC (273) vs. Credit Card (1,501), COD (365) vs. Cash on Delivery (149) 
- PreferedOrderCat	Mobile Phone (1,271) and Mobile (809) are separated



- Churn Rate: 16.8% overall customer churn rate (Churn = 1) -> class imbalance
- Complaint Rate: 28.5% of customers registered a complaint in the last month. primary driver for churn and unnecessary CO₂ emissions from returns.
- Customer Tenure: Averages 10.2 months (ranging from brand-new customers at 0 months to loyal customers at 61 months). Churn is heavily concentrated in early customer life ($\le 3$ months)
- Order Behavior: Customers average 3 orders per month, with an average cashback reward of $177.22 and an app usage average of ~2.9 hours
- Logistics & Orders:** Avg. 3 orders/month ($177.22 avg. cashback)
- WarehouseToHome spikes up to 127 km (outliers)

**Handling Inconsistency**

In [49]:
#standardise inconsistencies
#login devices
df['PreferredLoginDevice'] = df['PreferredLoginDevice'].replace({
   'Mobile Phone': 'Mobile',
   'Phone': 'Mobile'})

#payment modes
df['PreferredPaymentMode'] = df['PreferredPaymentMode'].replace({
   'CC': 'Credit Card',
   'COD': 'Cash on Delivery'})

#order categories
df['PreferedOrderCat'] = df['PreferedOrderCat'].replace({
   'Mobile Phone': 'Mobile'})

In [51]:
#cross check inconsistency
categorical_cols = df.select_dtypes(exclude=['number']).columns

for col in categorical_cols:
    display(df[col].value_counts())

PreferredLoginDevice
Mobile      3996
Computer    1634
Name: count, dtype: int64

PreferredPaymentMode
Debit Card          2314
Credit Card         1774
E wallet             614
Cash on Delivery     514
UPI                  414
Name: count, dtype: int64

Gender
Male      3384
Female    2246
Name: count, dtype: int64

PreferedOrderCat
Mobile                2080
Laptop & Accessory    2050
Fashion                826
Grocery                410
Others                 264
Name: count, dtype: int64

MaritalStatus
Married     2986
Single      1796
Divorced     848
Name: count, dtype: int64

**Imputing Data**

In [ ]:
#imputing missing values 
num_cols_with_na = [
    'Tenure', 'WarehouseToHome', 'HourSpendOnApp', 
    'OrderAmountHikeFromlastYear', 'CouponUsed', 
    'OrderCount', 'DaySinceLastOrder'
]

#grouped imputation by logical categories
df['WarehouseToHome'] = df['WarehouseToHome'].fillna(
    df.groupby('CityTier')['WarehouseToHome'].transform('median'))

df['HourSpendOnApp'] = df['HourSpendOnApp'].fillna(
    df.groupby('PreferredLoginDevice')['HourSpendOnApp'].transform('median'))

df['Tenure'] = df['Tenure'].fillna(
    df.groupby('PreferedOrderCat')['Tenure'].transform('median'))

df['OrderCount'] = df['OrderCount'].fillna(
    df.groupby('PreferedOrderCat')['OrderCount'].transform('median'))

#fallback: fill any remaining missing values with overall column medians
for col in num_cols_with_na:
    df[col] = df[col].fillna(df[col].median())

#cross check 0 missing values remain
print('missing values:', df[num_cols_with_na].isnull().sum().sum())

Remaining missing values: 0


In [ ]:
#cross check missing values
missing = df.isnull().sum()
missing[missing > 0]

Series([], dtype: int64)

**Handling Data Types**

In [54]:
#check dtypes
df.dtypes

CustomerID                       int64
Churn                            int64
Tenure                         float64
PreferredLoginDevice               str
CityTier                         int64
WarehouseToHome                float64
PreferredPaymentMode               str
Gender                             str
HourSpendOnApp                 float64
NumberOfDeviceRegistered         int64
PreferedOrderCat                   str
SatisfactionScore                int64
MaritalStatus                      str
NumberOfAddress                  int64
Complain                         int64
OrderAmountHikeFromlastYear    float64
CouponUsed                     float64
OrderCount                     float64
DaySinceLastOrder              float64
CashbackAmount                 float64
dtype: object

In [58]:
#convert count columns back to integers
count_cols = ['Tenure', 'WarehouseToHome', 'HourSpendOnApp', 
              'OrderAmountHikeFromlastYear', 'CouponUsed', 
              'OrderCount', 'DaySinceLastOrder']
df[count_cols] = df[count_cols].astype(int)

#convert CustomerID to string
df['CustomerID'] = df['CustomerID'].astype(str)

#convert categorical variables to 'category' dtype
cat_cols = ['PreferredLoginDevice', 'PreferredPaymentMode', 'Gender', 
            'PreferedOrderCat', 'MaritalStatus', 'CityTier']
df[cat_cols] = df[cat_cols].astype('category')

In [59]:
#cross check updated data types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype   
---  ------                       --------------  -----   
 0   CustomerID                   5630 non-null   str     
 1   Churn                        5630 non-null   int64   
 2   Tenure                       5630 non-null   int64   
 3   PreferredLoginDevice         5630 non-null   category
 4   CityTier                     5630 non-null   category
 5   WarehouseToHome              5630 non-null   int64   
 6   PreferredPaymentMode         5630 non-null   category
 7   Gender                       5630 non-null   category
 8   HourSpendOnApp               5630 non-null   int64   
 9   NumberOfDeviceRegistered     5630 non-null   int64   
 10  PreferedOrderCat             5630 non-null   category
 11  SatisfactionScore            5630 non-null   int64   
 12  MaritalStatus                5630 non-null   category
 13  NumberOfAddres

**Feature Engineering**

In [ ]:
#ustainability & returns proxies
df['Estimated_Returns'] = np.where(df['Complain'] == 1, 
                          np.ceil(df['OrderCount'] * 0.20), 
                          np.ceil(df['OrderCount'] * 0.05))
df['Estimated_CO2_kg'] = df['Estimated_Returns'] * 1.2

In [ ]:
#tenure buckets (high-risk window vs retained)
df['TenureBucket'] = pd.cut(df['Tenure'], bins=[-1, 3, 12, 60, 100], 
                     labels=['0-3 Mos (High Risk)', '4-12 Mos', '1-5 Yrs', '5+ Yrs'])

#warehouse distance category (logistics & sustainability risk)
df['DistanceCategory'] = pd.cut(df['WarehouseToHome'], bins=[-1, 10, 20, 50, 200], 
                         labels=['Close (0-10km)', 'Moderate (11-20km)', 'Far (21-50km)', 'Very Far (>50km)'])

In [63]:
#cross check feature outputs and check for missing values
print(df[['Estimated_Returns', 'Estimated_CO2_kg', 'TenureBucket', 'DistanceCategory']].isnull().sum())

Estimated_Returns    0
Estimated_CO2_kg     0
TenureBucket         0
DistanceCategory     0
dtype: int64


In [64]:
#import os
import os

#save cleaned dataset
os.makedirs('../data/processed', exist_ok=True)
df.to_csv('../data/processed/clean_data.csv', index=False)